Display project details (#182)
pupilfirst/wd301-scripts: handling-advanced-ui-events/display-project-details/README.md

# Text

> Now that we have added routes to display project, details, let's add a hyper link to the project name. So when a user clicks on it, they will be taken to detail page of the corresponding project.

Reviewer: ```suggestion
Now that we have added routes to display project, details, let's add a hyperlink to the project name. So when a user clicks on it, they will be taken to detail page of the corresponding project.
```

Open `src/pages/projects/ProjectListItems.tsx` in VS Code.

To create a hyperlink, we will use `Link` component from `react-router-dom` package. It will essentially be rendered as `anchor` tag in final html.

Let's import it first.

```tsx
import { Link } from "react-router-dom";
```

> `Link` component takes a `to` prop, which will be the `href` of final `a` tag that will be rendered. We will pass `id` as the value of `to` prop. Then the final ur will be a relative url which will be `/account/projects/:projectID`. Whenever we are creating a list of items in React, we will have to pass a `key` prop also. It is with value of the `key`, react uniquely identifies a component in a tree.

Reviewer: ```suggestion
`Link` component takes a `to` prop, which will be the `href` of final `a` tag that will be rendered. We will pass `id` as the value of `to` prop. Then the final url will be a relative url which will be `/account/projects/:projectID`. Whenever we are creating a list of items in React, we will have to pass a `key` prop also. It is with value of the `key`, react uniquely identifies a component in a tree.
```

> We will wrap the `h5` component to make it a hyperlink. It will be linked to `id` of the project.

Reviewer: ```suggestion
We will wrap the `h5` component to make it a hyperlink. It will be linked to `id` of the project. This is how `src/pages/projects/ProjectListItems.tsx` will look like.
```

  return (
    <>
      {projects.map((project: any) => (
        <Link
          key={project.id}
          to={`${project.id}`}
          className="block p-6 bg-white border border-gray-200 rounded-lg shadow hover:bg-gray-100 dark:bg-gray-800 dark:border-gray-700 dark:hover:bg-gray-700"
        >
          <h5 className="mb-2 text-xl font-medium tracking-tight text-gray-900 dark:text-white">
            {project.name}
          </h5>
        </Link>
      ))}
    </>
  );
}
```

Save the file. Now if you hover over the projects, you will see it is a hyperlink. If you click on it, it will render the `Show project details`.

Let's actually render some meaningful information about the project.

Let's create a folder `src/pages/project_details`. Inside the folder, let's create a file `ProjectDetails.tsx`.

In this component, we will extract the value of `projectID`, then filter it out from list of available projects to make sure it is a valid `id`. Then we will go on displaying the name of the project as a header.

`react-router-dom` provides a hook `useParams` to extract value of a [`dynamic segment`](https://reactrouter.com/en/main/route/route#dynamic-segments) in a url pattern.

> Add the following content to `ProjectDetails.tsx`

Author: @avishekjana Wouldn't renaming the component be enough? Every other pages has an `index.tsx` which is being used for rendering the root component of the url. (eg. logout, members, projects etc..) If the component is being renamed, then the import statement can be changed to:

```tsx
import ProjectDetailsContainer from "../pages/project_details/";

...

{
  path: "projects",
  element: <ProjectContainer />,
  children: [
    { index: true, element: <Projects /> },
    {
      path: ":projectID",
      element: <ProjectDetailsContainer />,
      children: [
        { index: true, element: <></> },

...
```
